src/circuit/boolean.rs

```diff
@@ -375,35 +375,44 @@ impl Boolean {
         where E: Engine,
               CS: ConstraintSystem<E>
     {
-        let c = Self::xor(&mut cs, a, b)?;
+        match (a, b) {
+            (&Boolean::Constant(a), &Boolean::Constant(b)) => {
+                if a == b {
+                    Ok(())
+                } else {
+                    Err(SynthesisError::Unsatisfiable)
+                }
+            },
+            (&Boolean::Constant(true), a) | (a, &Boolean::Constant(true)) => {
+                cs.enforce(
+                    || "enforce equal to one",
+                    |lc| lc + CS::one(),
+                    |lc| lc + CS::one(),
+                    |_| a.lc(CS::one(), E::Fr::one())
+                );
 
-        match c {
-            Boolean::Constant(false) => {
                 Ok(())
             },
-            Boolean::Constant(true) => {
-                Err(SynthesisError::Unsatisfiable)
-            },
-            Boolean::Is(ref res) => {
+            (&Boolean::Constant(false), a) | (a, &Boolean::Constant(false)) => {
                 cs.enforce(
-                    || "enforce equals zero",
+                    || "enforce equal to zero",
                     |lc| lc,
                     |lc| lc,
-                    |lc| lc + res.get_variable()
+                    |_| a.lc(CS::one(), E::Fr::one())
                 );
 
                 Ok(())
             },
-            Boolean::Not(ref res) => {
+            (a, b) => {
                 cs.enforce(
-                    || "enforce equals one",
-                    |lc| lc,
-                    |lc| lc,
-                    |lc| lc + CS::one() - res.get_variable()
+                    || "enforce equal",
+                    |lc| lc + CS::one(),
```

Comment:
Hehe. So, it's strictly better to have "one" be in the B term of the constraint rather than the A term, because the first operand `a` may not be directly boolean constrained (doesn't yet or ever appear in the B term), so it has some performance benefits (reduces density of polynomial queries in G2).

Reply:
It's an annoying problem that deserves a general fix at the bellman-level, really. Requires implementing some kind of SAT solver.